torch.compile documentation updates for 2024.2 (#24803)
openvinotoolkit/openvino: docs/articles_en/openvino-workflow/torch-compile.rst

How to Use
####################

To use ``torch.compile``, you need to define the ``openvino`` backend in your PyTorch application.
This way Torch FX subgraphs will be directly converted to OpenVINO representation without
any additional PyTorch-based tracing/scripting.

> .. code-block:: sh

Reviewer: ```suggestion
.. code-block:: python
```

   ...
   model = torch.compile(model, backend='openvino')
   ...

For OpenVINO versions below 2024.1, an additional import was needed to use ``openvino`` backend. However, OpenVINO package is now configured with `torch_dynamo_backends entrypoint <https://pytorch.org/docs/stable/torch.compiler_custom_backends.html#registering-custom-backends>`__. The additional import is not needed, if OpenVINO is installed as a pip package. For other install channels such as conda, import statement can be used as below.

> .. code-block:: sh

Reviewer: ```suggestion
.. code-block:: python
```

> For OpenVINO versions below 2024.1, an additional import was needed to use ``openvino`` backend. However, OpenVINO package is now configured with `torch_dynamo_backends entrypoint <https://pytorch.org/docs/stable/torch.compiler_custom_backends.html#registering-custom-backends>`__. The additional import is not needed, if OpenVINO is installed as a pip package. For other install channels such as conda, import statement can be used as below.

Reviewer 2: ```suggestion
This approach works only for the package distributed via pip, as it is now configured with
`torch_dynamo_backends entrypoint <https://pytorch.org/docs/stable/torch.compiler_custom_backends.html#registering-custom-backends>`__.

For OpenVINO installed via other channels, such as conda, and versions older than 2024.1, 
an additional import statement is needed:
```

How to Use
####################

To use ``torch.compile``, you need to define the ``openvino`` backend in your PyTorch application.
This way Torch FX subgraphs will be directly converted to OpenVINO representation without
any additional PyTorch-based tracing/scripting.
This approach works only for the **package distributed via pip**, as it is now configured with
`torch_dynamo_backends entrypoint <https://pytorch.org/docs/stable/torch.compiler_custom_backends.html#registering-custom-backends>`__.

.. code-block:: python

   ...
   model = torch.compile(model, backend='openvino')
   ...

For OpenVINO installed via channels other than pip, such as conda, and versions older than
2024.1, an additional import statement is needed:

.. code-block:: python

   import openvino.torch

   ...
   model = torch.compile(model, backend='openvino')
   ...

> Execution diagram:

Reviewer 2: ```suggestion
```

How to Use
####################

To use ``torch.compile``, you need to define the ``openvino`` backend in your PyTorch application.
This way Torch FX subgraphs will be directly converted to OpenVINO representation without
any additional PyTorch-based tracing/scripting.

> 

Reviewer 2: ```suggestion
This approach works only for the **package distributed via pip**, as it is now configured with
`torch_dynamo_backends entrypoint <https://pytorch.org/docs/stable/torch.compiler_custom_backends.html#registering-custom-backends>`__.
```

How to Use
####################

To use ``torch.compile``, you need to define the ``openvino`` backend in your PyTorch application.
This way Torch FX subgraphs will be directly converted to OpenVINO representation without
any additional PyTorch-based tracing/scripting.

.. code-block:: python

   ...
   model = torch.compile(model, backend='openvino')
   ...

This approach works only for the package distributed via pip, as it is now configured with
`torch_dynamo_backends entrypoint <https://pytorch.org/docs/stable/torch.compiler_custom_backends.html#registering-custom-backends>`__.

> 
> For OpenVINO installed via other channels, such as conda, and versions older than 2024.1, 

Reviewer 2: ```suggestion
For OpenVINO installed via other channels, such as conda, and versions older than 2024.1, 
```

> This approach works only for the package distributed via pip, as it is now configured with
> `torch_dynamo_backends entrypoint <https://pytorch.org/docs/stable/torch.compiler_custom_backends.html#registering-custom-backends>`__.

Reviewer 2: ```suggestion
```

> This approach works only for the package distributed via pip, as it is now configured with
> `torch_dynamo_backends entrypoint <https://pytorch.org/docs/stable/torch.compiler_custom_backends.html#registering-custom-backends>`__.
> For OpenVINO installed via other channels, such as conda, and versions older than 2024.1, 
> an additional import statement is needed:

Reviewer 2: ```suggestion
For OpenVINO installed via channels other than pip, such as conda, and versions older than
2024.1, an additional import statement is needed:
```

How to Use
####################

To use ``torch.compile``, you need to define the ``openvino`` backend in your PyTorch application.
This way Torch FX subgraphs will be directly converted to OpenVINO representation without
any additional PyTorch-based tracing/scripting.
This approach works only for the **package distributed via pip**, as it is now configured with
`torch_dynamo_backends entrypoint <https://pytorch.org/docs/stable/torch.compiler_custom_backends.html#registering-custom-backends>`__.

.. code-block:: python

   ...
   model = torch.compile(model, backend='openvino')
   ...

For OpenVINO installed via channels other than pip, such as conda, and versions older than
2024.1, an additional import statement is needed:

.. code-block:: python

   import openvino.torch

   ...
   model = torch.compile(model, backend='openvino')
   ...

> .. image:: ../assets/images/torch_compile_backend_openvino.svg

Reviewer 2: ```suggestion
.. image:: ../assets/images/torch_compile_backend_openvino.svg
   :alt: torch.compile execution diagram 
```

Windows support
+++++++++++++++++++++

PyTorch supports ``torch.compile`` officially on Windows from version 2.3.0 onwards.

> For PyTorch versions below 2.3.0, ``torch.compile`` feature is not supported on Windows officially. However, it can be accessed by running
> the below instructions:

Reviewer 2: ```suggestion
For PyTorch versions below 2.3.0, the ``torch.compile`` feature is not supported on Windows
officially. However, it can be accessed by running the following instructions:
```

PyTorch 2 export quantization is supported by OpenVINO backend in ``torch.compile``. To be able to access this feature, please follow the steps provided in `PyTorch 2 Export Post Training Quantization with X86 Backend through Inductor <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_x86_inductor.html>`__ and update the provided sample as explained below.

1. If you are using the PyTorch version 2.3.0 or above, disable constant folding in quantization to be able to benefit from the optimization in OpenVINO backend. This can be done by passing ``fold_quantize=False`` parameter into the ``convert_pt2e`` function.

    Update this line below:

   .. code-block:: python

      converted_model = convert_pt2e(prepared_model)

    As below:

   .. code-block:: python

      converted_model = convert_pt2e(prepared_model, fold_quantize=False)

2. Set ``torch.compile`` backend as OpenVINO and execute the model.

    Update this line below:

   .. code-block:: python

      optimized_model = torch.compile(converted_model)

    As below:

   .. code-block:: python

      optimized_model = torch.compile(converted_model, backend="openvino")

> Torchserve Integration

Reviewer: ```suggestion
TorchServe Integration
```

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

> Torchserve is a performant, flexible and easy to  use tool for serving Pytorch models in production. For more information on the details of Torchserve,
> you can refer to `Torchserve github repository. <https://github.com/pytorch/serve>`__. With OpenVINO ``torch.compile`` integration into Torchserve you can serve
> Pytorch models in production and accelerate them with OpenVINO on various Intel hardware. Detailed instructions on how to use OpenVINO with Torchserve are
> available in `Torchserve examples. <https://github.com/pytorch/serve/tree/master/examples/pt2/torch_compile_openvino>`__

Reviewer: ```suggestion
TorchServe is a performant, flexible and easy to  use tool for serving PyTorch models in production. For more information on the details of TorchServe,
you can refer to `TorchServe github repository. <https://github.com/pytorch/serve>`__. With OpenVINO ``torch.compile`` integration into TorchServe you can serve
PyTorch models in production and accelerate them with OpenVINO on various Intel hardware. Detailed instructions on how to use OpenVINO with TorchServe are
available in `TorchServe examples. <https://github.com/pytorch/serve/tree/master/examples/pt2/torch_compile_openvino>`__
```

       if sys.version_info >= (3, 11):
           `raise RuntimeError("Python 3.11+ not yet supported for torch.compile")

Support for PyTorch 2 export quantization (Preview)
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

PyTorch 2 export quantization is supported by OpenVINO backend in ``torch.compile``. To be able
to access this feature, follow the steps provided in
`PyTorch 2 Export Post Training Quantization with X86 Backend through Inductor <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_x86_inductor.html>`__
and update the provided sample as explained below.

> 1. If you are using the PyTorch version 2.3.0 or above, disable constant folding in quantization to be able to benefit from the optimization in OpenVINO backend. This can be done by passing ``fold_quantize=False`` parameter into the ``convert_pt2e`` function.
> 
>     Update this line below:
> 
>    .. code-block:: python
> 
>       converted_model = convert_pt2e(prepared_model)
> 
>     As below:
> 
>    .. code-block:: python
> 
>       converted_model = convert_pt2e(prepared_model, fold_quantize=False)
> 
> 2. Set ``torch.compile`` backend as OpenVINO and execute the model.
> 
>     Update this line below:
> 
>    .. code-block:: python
> 
>       optimized_model = torch.compile(converted_model)
> 
>     As below:
> 
>    .. code-block:: python
> 
>       optimized_model = torch.compile(converted_model, backend="openvino")

Reviewer 2: ```suggestion
1. If you are using PyTorch version 2.3.0 or later, disable constant folding in quantization to
   be able to benefit from the optimization in the OpenVINO backend. This can be done by passing
   ``fold_quantize=False`` parameter into the ``convert_pt2e`` function. To do so, change this
   line:

   .. code-block:: python

      converted_model = convert_pt2e(prepared_model)

   to the following:

   .. code-block:: python

      converted_model = convert_pt2e(prepared_model, fold_quantize=False)

2. Set ``torch.compile`` backend as OpenVINO and execute the model.

   Update this line below:

   .. code-block:: python

      optimized_model = torch.compile(converted_model)

   As below:

   .. code-block:: python

      optimized_model = torch.compile(converted_model, backend="openvino")
```

to access this feature, follow the steps provided in
`PyTorch 2 Export Post Training Quantization with X86 Backend through Inductor <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_x86_inductor.html>`__
and update the provided sample as explained below.

1. If you are using PyTorch version 2.3.0 or later, disable constant folding in quantization to
   be able to benefit from the optimization in the OpenVINO backend. This can be done by passing
   ``fold_quantize=False`` parameter into the ``convert_pt2e`` function. To do so, change this
   line:

   .. code-block:: python

      converted_model = convert_pt2e(prepared_model)

   to the following:

   .. code-block:: python

      converted_model = convert_pt2e(prepared_model, fold_quantize=False)

2. Set ``torch.compile`` backend as OpenVINO and execute the model.

   Update this line below:

   .. code-block:: python

      optimized_model = torch.compile(converted_model)

   As below:

   .. code-block:: python

      optimized_model = torch.compile(converted_model, backend="openvino")

TorchServe Integration
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

> TorchServe is a performant, flexible and easy to  use tool for serving PyTorch models in production. For more information on the details of TorchServe,

Reviewer 2: ```suggestion

TorchServe is a performant, flexible, and easy to use tool for serving PyTorch models in production. For more information on the details of TorchServe,
```

       if sys.version_info >= (3, 11):
           `raise RuntimeError("Python 3.11+ not yet supported for torch.compile")

Support for PyTorch 2 export quantization (Preview)
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

> PyTorch 2 export quantization is supported by OpenVINO backend in ``torch.compile``. To be able to access this feature, please follow the steps provided in `PyTorch 2 Export Post Training Quantization with X86 Backend through Inductor <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_x86_inductor.html>`__ and update the provided sample as explained below.

Reviewer 2: ```suggestion
PyTorch 2 export quantization is supported by OpenVINO backend in ``torch.compile``. To be able
to access this feature, follow the steps provided in
`PyTorch 2 Export Post Training Quantization with X86 Backend through Inductor <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq_x86_inductor.html>`__
and update the provided sample as explained below.
```
